blockprotocol/blockprotocol
# [SITE] Schema editor frontend

This PR introduces a basic schema editor to the site. It's mostly lifted from the one in [HASH](https://github.com/hashintel/hash/tree/main/packages/hash/frontend/src/components/entityTypes/SchemaEditor) with a few tweaks:
1. Use MUI components in this repo
2. Disable sub-schema support (unsure if we want to encourage defining things as sub-schemas rather than separate entities)
3. New wrapping page + plumbing with the API
4. Add ability to link a property field to a schema.org property

**Suggestion for reviewers**
The components in `site/src/components/entityTypes/SchemaEditor/` are basically the same as the ones that have been already reviewed elsewhere. I'd focus on new functionality:
- `updatePropertySchemaOrgLink`
- the wrapping page at `pages/[shortname]/types/[title].tsx`
- the canonical URI at `pages/types/[entityTypeId].tsx` which redirects to the one above, unless you add `?json` or request it with an  `accepts: "application/json"` header, in which case you'll get the raw JSON
- changes t `site/src/lib/jsonSchema.ts`

**To test**
- Create an entity type via the API following instructions in the README or #128 
- visit `/types/[entityTypeId]`, you should be sent to the form
- amend some stuff and refresh to check that changes persist correctly
- visit the raw json at `/types/[entityTypeId]?json` and check it makes sense

**Known issues**
- There is some unused code related to sub-schemas I need to clear up afterwards.
- There's a patch for `ajv` in `site/patches` rather than at the workspace root: running `patch-package` from the root does not generate a patch, I think because the edited code is in `site/node_modules`. I might be doing something wrong with the library, unsure.

**Screenshots**

![Screenshot 2022-01-26 at 18 51 35](https://user-images.githubusercontent.com/37743469/151236452-ac99e4be-2fff-4285-93e5-952a940b09a9.png)

---

![Screenshot 2022-01-26 at 18 51 40](https://user-images.githubusercontent.com/37743469/151236459-48f38c16-7b79-469f-9c60-0bac369e1c5d.png)
